patches/api/0404-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,1140 @@
+@@ -269,5 +269,12 @@ public interface UnsafeValues {
+      * @throws IllegalStateException if no biome by the given key is registered.
+      */
+     void setBiomeKey(RegionAccessor accessor, int x, int y, int z, NamespacedKey biomeKey);
++
++    /**
++     * Gets an instance of a static damage origin from its key
++     * @param key the key
++     * @return the damage origin
++     */
++    io.papermc.paper.entity.damage.DamageOrigin getDamageOrigin(@org.jetbrains.annotations.NotNull String key);
+     // Paper end
+ }
+diff --git a/src/main/java/org/bukkit/entity/Damageable.java b/src/main/java/org/bukkit/entity/Damageable.java
+index fc4d3bcd9b16097086fef7975274d825b65adb10..99bef2d9b5fa45e14b247189a29365cc966789c3 100644
+--- a/src/main/java/org/bukkit/entity/Damageable.java
++++ b/src/main/java/org/bukkit/entity/Damageable.java
+@@ -8,7 +8,7 @@ import org.jetbrains.annotations.Nullable;
+  */
+ public interface Damageable extends Entity {
+     /**
+-     * Deals the given amount of damage to this entity.
++     * Deals the given amount of damage to this entity using the {@link io.papermc.paper.entity.damage.VanillaDamageOrigin#GENERIC} damage origin. <!-- Paper Damage source wrapper -->
+      *
+      * @param amount Amount of damage to deal
+      */
+@@ -20,9 +20,64 @@ public interface Damageable extends Entity {
+      *
+      * @param amount Amount of damage to deal
+      * @param source Entity which to attribute this damage from
++     * @deprecated you should use {@link #damage(double, io.papermc.paper.entity.damage.DamageOrigin)} using a {@link io.papermc.paper.entity.damage.EntityDamageOrigin EntityDamageOrigin} instead <!-- Paper Damage source wrapper -->
+      */
++    @Deprecated // Paper - Damage source wrapper
+     void damage(double amount, @Nullable Entity source);
+ 
++    // Paper start - Damage source wrapper
++    /**
++     * Deals the given amount of damage to this entity with specific {@link io.papermc.paper.entity.damage.DamageOrigin DamageOrigin}.
++     *
++     * <p>
++     * In order to anticipate if this damage will deal some damage you can use the
++     * {@link org.bukkit.entity.Entity#isInvulnerableTo(io.papermc.paper.entity.damage.DamageOrigin) Entity#isInvulnerableTo} is false.
++     * <br>
++     * <p>
++     * Bukkit has linked the block within the damage source in order to detect
++     * the {@link org.bukkit.event.entity.EntityDamageByBlockEvent EntityDamageByBlockEvent} that's why
++     * you should use {@link #damageByBlock(double, io.papermc.paper.entity.damage.DamageOrigin, org.bukkit.block.Block)}.
++     *
++     * @param amount Amount of damage
++     * @param origin Origin of damage
++     *
++     * @return Whether the damage has been dealt
++     * @see #damageByBlock(double, io.papermc.paper.entity.damage.DamageOrigin, org.bukkit.block.Block)
++     * @throws IllegalArgumentException If this association isn't bukkit compatible for the damage events
```

Comment:
you'll want to elaborate more here what "association" means here.

Reply:
This is intended, see the impl of this temp check but basically i want to change this logic to bypass this logic while keeping the same behavior for plugins event after the hardfork. The entity will be now the direct source/source depending of the event, the getEntity will be deprecated in favor of the damage origin entities. The static field entityDamage will be deleted and the blockDamage will be a parameter in the CraftEventFactory to avoid concurrency issue. And finally the DamageModifier will change also a little bit.